runtime/common/src/precompile/schedule_call.rs

```diff
@@ -155,37 +161,97 @@ impl<AccountId, AddressMapping, Scheduler, ChargeTransactionPayment, Call, Origi
 				)
 				.into();
 
-				let delay = DispatchTime::After(min_delay);
-				let origin = Origin::root().caller().clone();
-
 				let current_id = EvmSchedulerNextID::get();
 				let next_id = current_id
 					.checked_add(1)
 					.ok_or_else(|| ExitError::Other("Scheduler next id overflow".into()))?;
 				EvmSchedulerNextID::set(&next_id);
 
-				let task_address = Scheduler::schedule_named(
-					Encode::encode(&(&"ScheduleCall", current_id)),
-					delay,
+				// task_id = id + sender + reserve_fee
+				let task_id = join_task_id(Encode::encode(&(&"ScheduleCall", current_id)), from, _fee.into());
+				Scheduler::schedule_named(
+					task_id.clone(),
+					DispatchTime::After(min_delay),
 					None,
 					0,
-					origin,
+					Origin::root().caller().clone(),
 					call,
 				)
-				.map_err(|_| ExitError::Other("Scheduler failed".into()))?;
+				.map_err(|_| ExitError::Other("Schedule failed".into()))?;
+
+				Ok((ExitSucceed::Returned, task_id, 0))
+			}
+			Action::Cancel => {
+				let from = input.evm_address_at(1)?;
+				let task_id = input.bytes_at(2 * PER_PARAM_BYTES, 96)?;
+
+				debug::debug!(
+					target: "evm",
+					"cancel call: from: {:?}, task_id: {:?}",
+					from,
+					task_id,
+				);
+
+				// task_id = id + sender + reserve_fee
+				let (_, sender, fee) = split_task_id(&task_id);
+				ensure!(sender == from, ExitError::Other("NoPermission".into()));
+
+				#[cfg(not(feature = "with-ethereum-compatibility"))]
+				{
+					// unreserve the transaction fee for gas_limit
+					let from_account = AddressMapping::get_account_id(&from);
+					ChargeTransactionPayment::unreserve_fee(&from_account, fee.into());
```

Comment:
this can only happen if cancel is successful. i.e. we have verified the task id is valid